Storybook: Support proper extraction of TypeScript prop types

mirka opened the pull request:
## Description

This PR clears the path for proper extraction of component prop types to use in our Storybook and documentation. Given our ongoing investment in TypeScript migrations, it would be great to leverage this as much as possible.

I tweaked two components here (`Heading` and `Divider`) to demonstrate what it could look like. A big benefit is that the prop type documentation **and** all Controls can be generated automatically for the most part, with inline descriptions ✨   In some cases we may want to override or hide specific controls, but other than that we get everything for free.

### Details

After a lot of experimentation, I figured out some things that were preventing Storybook's `react-docgen-typescript` from properly extracting props information from our TypeScript/JSDocs. Here's what's required for proper type extraction:

- The component has to be written in actual TypeScript, not just typed with a JSDoc comment 😅 
- The `WordPressComponentProps` type alias needs to be imported from a proper `.ts` source file. The problem was that our components were importing it via `ui/context/index.js`, which is not a `.ts` file.
- The `contextConnect()` HOC needs to be in TypeScript.
- The connected component needs to be exported individually as a named export, not just as `default`.

### What this means for TypeScript migration and story authoring

As more components are converted to TypeScript and their stories are improved, we can expect that **most consumer devs will want to interact with Storybook primarily through the Docs view**, rather than the Canvas. Especially with code samples (92354c96a3f1e8a698284a981bf2ffb05fbb3d62) visible for each story, it's much easier to understand how to use the components. And for the default story, the Controls are available right in the Docs view. Canvas view will only be necessary for advanced debugging.

Eventually, I imagine we'll want to make the [Docs mode version](https://storybook.js.org/docs/react/writing-docs/build-documentation) of our Storybook to be our canonical documentation, instead of the limited [Component Reference](https://developer.wordpress.org/block-editor/reference-guides/components/) currently on the dotorg site.

So whenever possible, it'll be beneficial for us to write TypeScript and Storybook stories in a way that makes sense for the Docs view. For example:

- Are there any weird props that show up in Docs view that you want to hide? Consider whether it should be `Omit`-ed at the TypeScript level. (Example: #38844)
- Are some stories showing unhelpful code snippets in Docs view? See if you can tweak the story to show an understandable code sample.

## Testing Instructions

1. `npm run storybook:dev`
2. See stories for `Heading` and `Divider`. Compare with the ones on https://wordpress.github.io/gutenberg/
3. If you're interested, try `npx start-storybook -c ./storybook -p 50240 --docs` to see how it would look like in Docs mode.

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/555336/154171498-07d10d0f-3f43-4509-b3cb-346205619cfa.mp4

## Types of changes

- Minor, non-consumer-facing changes (Changing files to `.ts`, re-exporting as named components)
- Storybook changes

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

mirka committed:
Export WordPressComponent types properly

mirka committed:
Storybook: Allow tsx stories

mirka committed:
Heading: Use unconnected component for extraction

Export unconnected `Heading`

mirka committed:
Divider: Extract types for stories

mirka committed:
Expand code by default in Docs tab

mirka committed:
Heading: Tweak story for Controls

mirka committed:
Divider: Tweak story for "all Controls" view

mirka committed:
Convert contextConnect to TS

mirka committed:
Heading: Simply named export connected component

mirka committed:
Heading: Export with correct name

This is required for the code snippets to show the correct component name.

mirka committed:
Heading: Fix doc comment

mirka committed:
Divider: Rejigger exports

mirka commented on packages/components/src/divider/component.tsx:
````diff
@@ -53,6 +53,6 @@ function Divider(
  * ```
  */
-const ConnectedDivider = contextConnect( Divider, 'Divider' );
+export const Divider = contextConnect( UnconnectedDivider, 'Divider' );
````
An interesting side effect of adding this named export is that, the `displayName` now becomes whatever this const is named. Since "connected" is an internal concern and Divider isn't even usable when unconnected, I think it would be better to name this `Divider` rather than `ConnectedDivider`.

ciampo replied:
I thought that the `displayName` was [assigned by `contextConnect`](https://github.com/WordPress/gutenberg/blob/811a2427b93a764206be2cf4ade101edd18d2608/packages/components/src/ui/context/context-connect.ts#L70) based on the namespace passed as the 2nd argument (which is also `Divider` in this case).

But yeah, regardless of that, I think the naming convention that you propose makes sense!

mirka replied:
> I thought that the `displayName` was [assigned by `contextConnect`](https://github.com/WordPress/gutenberg/blob/811a2427b93a764206be2cf4ade101edd18d2608/packages/components/src/ui/context/context-connect.ts#L70) based on the namespace passed as the 2nd argument (which is also `Divider` in this case).

It's surprising to me too! It must be some kind of webpack thing...? You can observe the same behavior in React devtools, not just in Storybook code snippets.

mirka committed:
Heading: Convert story to TS

mirka committed:
Divider: Convert story to TS

mirka commented on packages/components/src/divider/stories/index.tsx:
```diff
@@ -23,18 +24,26 @@ export default {
+		controls: { expanded: true },
+	},
+} as ComponentMeta< typeof Divider >;
 
-const HorizontalTemplate = ( args ) => (
+const HorizontalTemplate: ComponentStory< typeof Divider > = ( args ) => (
```
These `Component*` typings add a bit to the boilerplate, but they are what prevent mistakes in the `argTypes` and `args` objects.

ciampo replied:
I think it's a necessary change, which hopefully gives better docs and stricter TS checks in general

mirka commented on packages/components/src/divider/stories/index.tsx:
```diff
@@ -45,20 +54,23 @@ const VerticalTemplate = ( args ) => {
 	return (
 		<div className={ wrapperClassName }>
```
Do you think we should now inline these styles (`style={{ display: 'flex' }}` etc.) so it's clearer in the code snippet (in Docs view) what's going on? I think our initial thinking was that we should demonstrate how to use Emotion in the examples, but on second thought it's probably irrelevant to consumers who don't use Emotion.

<img src="https://user-images.githubusercontent.com/555336/154546191-d076df30-8613-4939-88b5-bb67615aff1b.png" alt="Code snippet in Docs view that is exposing an Emotion classname" width="300">

ciampo replied:
This is an interesting one.

On one end, it's good to show the intended use of Emotion, on the other it makes Storybook examples harder to read — and in a way, it "exposes" Emotion (which should be an internal dependency) to the package's consumers.

We could probably use inline styles almost everywhere, except for a few stories where using Emotion is the point of the story (e.g. the `useCx` stories)

mirka replied:
Agreed. I went ahead and inlined it for this one ([a7181bf](https://github.com/WordPress/gutenberg/pull/38842/commits/a7181bfbdca9cee3a399e3fc6886f535d4c1f95f)).

ciampo commented on packages/components/src/divider/stories/index.tsx:
```diff
@@ -23,18 +24,26 @@ export default {
 			control: { type: 'number' },
 		},
+		// TODO: This is a reakit internal prop. We can hide it from the props table like this
+		// if we don't want to expose it, but should we rather omit it at the TypeScript level?
```
> should we rather omit it at the TypeScript level?

I'd say so, especially since it's `unstable`

mirka replied:
Done in [ee1922b](https://github.com/WordPress/gutenberg/pull/38842/commits/ee1922b814093634810392c58cf57caf0d45f532)

ciampo reviewed (commented):
Thank you @mirka , seeing this exploratory work taking shape makes very excited!

> A big benefit is that the prop type documentation and all Controls can be generated automatically for the most part, with inline descriptions ✨ In some cases we may want to override or hide specific controls, but other than that we get everything for free.

This is great! It will interesting to see, as we experiment with this, which other limitations we're going to hit. In that sense, it's going to be very important for us to choose the docgen "engine" that fits our needs the most.

> After a lot of experimentation, I figured out some things that were preventing

This part was very interesting to read, but it all made sense to me. This also legitimises the priority that we should give to #35744 

> we can expect that most consumer devs will want to interact with Storybook primarily through the Docs view

Yes 🚀 

> Eventually, I imagine we'll want to make the [Docs mode version](https://storybook.js.org/docs/react/writing-docs/build-documentation) of our Storybook to be our canonical documentation, instead of the limited [Component Reference](https://developer.wordpress.org/block-editor/reference-guides/components/) currently on the dotorg site.
> ...
> So whenever possible, it'll be beneficial for us to write TypeScript and Storybook stories in a way that makes sense for the Docs view.

That's what I have in mind, too.

> Are there any weird props that show up in Docs view that you want to hide? Consider whether it should be Omit-ed at the TypeScript level. (Example: https://github.com/WordPress/gutenberg/pull/38844)

Specifically on the `as` prop, I think it'd be good to expose it, since some components are polymorphic, while others are not. WDYT ?

> Are some stories showing unhelpful code snippets in Docs view? See if you can tweak the story to show an understandable code sample.

This will probably need a story-by-story "human" review.

Maybe a good balance for our story examples could be:
- Default story: expose as many props as possible, so that a dev can tweak them while reading the docs
- Non-default stories: more specific example/scenarios, with very few controls exposed


How much work is left in order to get this extraction to a "good enough" point? Woudl it require all components and Storybook examples to be migrated to TypeScript first?

mirka committed:
Replace Emotion with inline style

mirka committed:
Move source expansion setting to per-component

mirka commented:
> it's going to be very important for us to choose the docgen "engine" that fits our needs the most.

Right. FWIW, out of the [two docgens](https://github.com/storybookjs/storybook/blob/next/addons/docs/react/README.md#typescript-props-with-react-docgen) that Storybook supports, `react-docgen` is a non-starter because it doesn’t support imported types 😬 And `react-docgen-typescript` seems to be the only remaining major option for React TS stuff, so we might not have much of a choice.

> Specifically on the `as` prop, I think it'd be good to expose it, since some components are polymorphic, while others are not. WDYT ?

Yes, it will be present for polymorphic components, and omitted from non-polymorphic components, if that’s what you’re asking 👍

> Maybe a good balance for our story examples could be:
> 
> * Default story: expose as many props as possible, so that a dev can tweak them while reading the docs
> * Non-default stories: more specific example/scenarios, with very few controls exposed

The thing is, all props recognized by the docgen will automatically be exposed as Controls on all stories by default. At that point, if we want to hide any Controls we have to disable each one specifically (denylist). This is how Storybook works, and it’s not something we can easily change to be an allowlist system. (The only reason we weren’t getting this behavior is because the majority of our types are currently unrecognizable by the docgens 😅)

So in most cases I don’t think it is worth the effort to carefully curate the set of Controls exposed on specific stories.

This might seem like an annoyance, but ultimately I think the behavior is sensible, and informs us how we should be approaching the different Storybook sections. The Docs view code snippet is for understanding which props are being explicitly set. The Canvas view is for advanced tinkering and debugging, in which case we actually _do_ want to expose as many Controls as possible.

Is that reasonable?

> How much work is left in order to get this extraction to a "good enough" point? Woudl it require all components and Storybook examples to be migrated to TypeScript first?

I think we can approach it in a “progressive enhancement” way. First off we can merge this PR, since it doesn’t have ill effects on any existing stories.  Then every time a TS conversion is complete, we can tweak the stories for that component so it makes use of the type data. Once the majority of TS conversions are complete, we can set the default tab to Docs. Once 100% of the TS conversions are complete, we can remove markdown docs and point everyone to the Storybook.

mirka committed:
Remove `unstable_system` prop from types

ciampo commented:
> out of the [two docgens](https://github.com/storybookjs/storybook/blob/next/addons/docs/react/README.md#typescript-props-with-react-docgen) that Storybook supports, `react-docgen` is a non-starter because it doesn’t support imported types 😬 And `react-docgen-typescript` seems to be the only remaining major option for React TS stuff, so we might not have much of a choice.

That's good to know. Out of curiosity, what does `react-docgen` do better than `react-docgen-typescript`? Are both projects currently well maintained and easy to contribute to? Just trying to make sure that picking a docgen now won't hurt us in the future.

> Yes, it will be present for polymorphic components, and omitted from non-polymorphic components, if that’s what you’re asking 👍

Yup, awesome!

> The thing is, all props recognized by the docgen will automatically be exposed as Controls on all stories by default. At that point, if we want to hide any Controls we have to disable each one specifically (denylist) ... So in most cases I don’t think it is worth the effort to carefully curate the set of Controls exposed on specific stories.
> ...
> The Docs view code snippet is for understanding which props are being explicitly set. The Canvas view is for advanced tinkering and debugging, in which case we actually _do_ want to expose as many Controls as possible.
> 
> Is that reasonable?

Yes, thank you for the explanation. I guess we should stil add multiple stories to the Canvas view showcasing specific scenarios / recipes?

> I think we can approach it in a “progressive enhancement” way. First off we can merge this PR, since it doesn’t have ill effects on any existing stories. Then every time a TS conversion is complete, we can tweak the stories for that component so it makes use of the type data. Once the majority of TS conversions are complete, we can set the default tab to Docs. Once 100% of the TS conversions are complete, we can remove markdown docs and point everyone to the Storybook.

Sounds like a plan to me. I will give this PR a final review then!

ciampo commented on packages/components/src/ui/context/context-connect.ts:
```diff
@@ -24,15 +30,16 @@ import { getStyledClassNameFromKey } from './get-styled-class-name-from-key';
-export function contextConnect( Component, namespace, options = {} ) {
-	/* eslint-enable jsdoc/valid-types */
+export function contextConnect< P >(
+	Component: ( props: P, ref: Ref< any > ) => JSX.Element | null,
```
should we use the `ForwardedRef` type here?

mirka replied:
Done ✅

ciampo commented on packages/components/src/heading/stories/index.tsx:
```diff
@@ -0,0 +1,28 @@
+		color: { control: { type: 'color' } },
+	},
+	parameters: {
+		controls: { expanded: true },
+		docs: { source: { state: 'open' } },
```
Would it make sense to make these parameters default on the Storybook global config, or is this something that should be done only after all components/stories have been migrated to TS ?

mirka replied:
Yeah, I started out by setting them globally, but I checked a couple stories and they were too noisy or meaningless 🤷

ciampo replied:
Makes sense!

ciampo reviewed (commented):
Apart from the previous comment and the inline comments, here's some additional thoughts:

1. Is there a way to **sort the props** displayed in the docs alphabetically? When there's a lot props, it may help to have a way of sorting them so that a user can easily find them (Ideally we could even group them, but it may be a task that we want to tackle separately)
2. A lot of props can't be expanded to their "raw" value (e.g the `adjustLineHeightForInnerControls` prop of `Heading` has a type of `TextAdjustLineHeightForInnerControls`). Although this is not great, the opposite approach (i.e expanding every alias) could be even worse. Probably **better for now to keep the alias expansion deactivated?**
3. When guessing a prop type for the associated control, S**torybook seems to default to `Object`**, which is often not the right choice (this happens, for example, for all CSS-related props) and can undermine the usefulness of the docs. Probably the only solution here is to manually override those control types for each story that gets converted?
4. As a special case of the previous point, **the `as` prop is also presented as an Object. Clicking on the "Set object" button currently crashes Storybook**
5. We should also look into adding a **description for the `as` prop** in the `WordPressComponent` typerybook defaults to an Object type, which breaks storybook

And finally, some additional feedback specific to the `Heading` and `Divider` stories:

- Heading: not something introduced by this PR, but exposing the component through the new story revealed quite **a few unpredictable (and very likely unintended) behaviors**, especially around tweaking character-based and line-based truncation. Also, the `display` prop value always gets overridden by a `display: block' rule declared later
- Divider: in the default "Horizontal" story (which is used to highlight all `Divider` props), the user is able to select `vertical` for the orientation, but **currently the story doesn't really support that use case** (i.e. the `Divider` can't be seen). We should probably:
    - Update the story by adding a flex wrapper
    - Update the docs of the `orientation` prop

mirka committed:
Text: Remove extraneous type alias

mirka committed:
Enforce TS checks on ComponentMeta argTypes

mirka committed:
Don't exclude tsx stories from type check

mirka committed:
Improve prop sorting

`requiredFirst` seems to also sort the rest of the props alphabetically 🎉

mirka committed:
Add more control type hints

mirka committed:
Add JSDoc description for `as` prop

mirka committed:
Change `Ref` to `ForwardedRef`

mirka commented on packages/components/tsconfig.json:
```diff
@@ -86,7 +86,7 @@
 		"src/**/*.native.js",
 		"src/**/react-native-*",
-		"src/**/stories",
+		"src/**/stories/**.js", // only exclude js files, tsx files should be checked
```
Included tsx stories in type checks ✅

ciampo replied:
Good catch!

mirka commented:
I think we're in a pretty good place, thanks to your feedback! All issues addressed.

> That's good to know. Out of curiosity, what does `react-docgen` do better than `react-docgen-typescript`? Are both projects currently well maintained and easy to contribute to? Just trying to make sure that picking a docgen now won't hurt us in the future.

I don’t have much information aside from the [comparison table](https://github.com/storybookjs/storybook/blob/next/addons/docs/react/README.md#typescript-props-with-react-docgen), but apparently it’s faster and has fewer bugs, albeit with lower quality type analysis? Both repos seem reasonably active.

> I guess we should stil add multiple stories to the Canvas view showcasing specific scenarios / recipes?

Exactly! When there are a lot of items being exposed in the props table, it’s important that we showcase the common usage examples with stories.


> 1. Is there a way to **sort the props** displayed in the docs alphabetically?

Done ✅ Required props first, then the rest are alphabetical.

> 2. A lot of props can't be expanded to their "raw" value (e.g the `adjustLineHeightForInnerControls` prop of `Heading` has a type of `TextAdjustLineHeightForInnerControls`). Although this is not great, the opposite approach (i.e expanding every alias) could be even worse. Probably **better for now to keep the alias expansion deactivated?**

Yes, also I think another way of approaching it in certain cases is to actually _not use a type alias_. Like in your example, [`TextAdjustLineHeightForInnerControls`](https://github.com/WordPress/gutenberg/blob/a556b5218b248bc049ad2bfe71cf1baff828a8d4/packages/components/src/text/types.ts#L11) does not really need to be an alias. Interestingly, the logic whether to expand or not seems to be the same as in IntelliSense, so overusing them makes it harder to understand in IDEs too.

| Before | After |
|--------|-------|
|<img src="https://user-images.githubusercontent.com/555336/155205309-b9074c82-1693-4eb2-8e28-fbe048e1e5b9.png" alt="type alias collapsed in VS Code IntelliSense hover" width="400">|<img src="https://user-images.githubusercontent.com/555336/155205302-762def89-d4d8-463f-8c09-28ac75afa291.png" alt="types expanded" width="400">|



> 3. When guessing a prop type for the associated control, S**torybook seems to default to `Object`**, which is often not the right choice (this happens, for example, for all CSS-related props) and can undermine the usefulness of the docs. Probably the only solution here is to manually override those control types for each story that gets converted?

Currently, yes. Let’s see how often it happens. FWIW, there is a already a [basic mechanism](https://storybook.js.org/docs/react/essentials/controls#custom-control-type-matchers) to regex-match and map props to certain controls. It doesn’t support our needs yet, but the [code](https://github.com/storybookjs/storybook/blob/next/lib/store/src/inferControls.ts) looks pretty straightforward so it might be a good place to contribute. (I went ahead and upvoted in an [upstream feature request](https://github.com/storybookjs/storybook/issues/16818#issuecomment-1047290568).)

> 4. As a special case of the previous point, **the `as` prop is also presented as an Object. Clicking on the "Set object" button currently crashes Storybook**

Fixed ✅

> 5. We should also look into adding a **description for the `as` prop** in the `WordPressComponent`

Done ✅

> 
> And finally, some additional feedback specific to the `Heading` and `Divider` stories:

- `Heading`: Moved follow-ups to the TS migration issue notes for the `Text` component.
- `Divider`: Will follow up after I try to add [`storybook-description-loader`](https://github.com/izhan/storybook-description-loader)

ciampo commented on packages/components/src/ui/context/wordpress-component.ts:
```diff
@@ -18,6 +18,7 @@ export type WordPressComponentProps<
 	Omit< React.ComponentPropsWithRef< T >, 'as' | keyof P | 'children' > &
 	( IsPolymorphic extends true
 		? {
+				/** The HTML element to render the component as. */
```
The `as` prop can be also used to render a component as another component (not just an HTML element) — should we mention it in the comment?

mirka replied:
Good point, fixed.

ciampo commented on packages/components/src/text/types.ts:
```diff
@@ -35,7 +28,12 @@ export interface Props extends TruncateProps {
 	 * Automatically calculate the appropriate line-height value for contents that render text and Control elements (e.g. `TextInput`).
 	 */
-	adjustLineHeightForInnerControls?: TextAdjustLineHeightForInnerControls;
+	adjustLineHeightForInnerControls?:
+		| boolean
+		| 'large'
+		| 'medium'
+		| 'small'
+		| 'xSmall';
```
Maybe the "limit" for expanding type aliases is if the types that are unioned are all of the same general type or not?

E.g. a union of just strings could be "easier" to expand for the docgen, while a union of strings and boolean may be kept un-expanded?

mirka replied:
I don't know 🤷

ciampo replied:
Me neither, but good to keep at the back of our minds for future work :)

ciampo commented on packages/components/src/heading/stories/index.tsx:
```diff
@@ -0,0 +1,37 @@
+		letterSpacing: { control: { type: 'text' } },
+		lineHeight: { control: { type: 'text' } },
+		optimizeReadabilityFor: { control: { type: 'color' } },
+		variant: { control: { type: 'radio' }, options: [ 'muted' ] },
```
Nit: using a radio with only one option, it's not possible to "deselect" that option, once selected

mirka replied:
Kind of like the "reset" problem in the editor sidebar! I actually think this is consistent with all controls (especially auto-generated ones), in that you can't really go back to the `undefined` state once you set a value. We could of course add an `undefined` option, but I think the overall design intent is to use the "Reset controls" button ↩️ in the top right.

I'm not sure how I feel about this yet! On one hand I don't want to establish a pattern of adding a lot of manual tweaks to auto-generated controls, because it increases maintenance burden. But in some cases it may be annoying to have to hit Reset each time. So... maybe worth it for heavily trafficked props like `variant` on `Button`?

ciampo replied:
Agreed. Let's leave it as-is (and rely on the reset button for now). I believe we are amongst the heavy users of this Storybook anyway, so we can make this change if we ever feel like it's really needed.

ciampo commented on packages/components/src/heading/stories/index.tsx:
```diff
@@ -0,0 +1,37 @@
+	component: Heading,
+	title: 'Components (Experimental)/Heading',
+	argTypes: {
+		adjustLineHeightForInnerControls: { control: { type: 'text' } },
```
`adjustLineHeightForInnerControls` can also be a `boolean` — should we add `true` and `false` to the list of possible values, maybe in a dropdown-like control?

mirka replied:
Mmm yeah, this one’s a bit overcomplicated too. Based on the [code](https://github.com/WordPress/gutenberg/blob/storybook/prop-type-extraction/packages/components/src/text/get-line-height.ts), `true` gives the same result as `'medium'`, and `false` gives the same result as `undefined`. Before we graduate `<Text>` from experimental status, I'd prefer removing the boolean types from this prop API. (Kind of similar to your `UnitControl` cleanup!)

I'll note this in the TS refactor tracking issue.

ciampo replied:
Just to close the circle, this was finally done in #54953 !

ciampo commented on packages/components/src/heading/stories/index.tsx:
```diff
@@ -0,0 +1,37 @@
+		as: { control: { type: 'text' } },
+		color: { control: { type: 'color' } },
+		display: { control: { type: 'text' } },
+		letterSpacing: { control: { type: 'text' } },
+		lineHeight: { control: { type: 'text' } },
```
These values can be both strings or numbers, is there a way to achieve that with controls?

mirka replied:
The `object` control can strictly cover both, since that's basically a JSON string input so it differentiates between `1` and `"1"`. Though in the vast majority of cases I think a normal `text` control would be sufficient for tinkering and better for UX (because people will have trouble recognizing that they need to quote their strings in a `object` control).

ciampo reviewed (approved):
Basically, a solid 👍 to all of your answers above 🚀 

I left a few inline comments, but I am preentively approving this PR as they are all small changes — feel free to merge after resolving them if you think so!

Also, 👍  re. the proposed follow-ups:
- Adding https://github.com/izhan/storybook-description-loader
- Iterating on `Divider`'s story (especially around the vertical orientation)

mirka committed:
Clarify `as` prop comment

im3dabasia commented:
Hi , 

Some I was going through an issue which was to refactor code to TS for code inside the components package. 

Since some work was done to refactor the `packages/components/src/context` in this PR, I thought it's best to ask here if we are looking to refactor this files JS files into TS.

https://github.com/WordPress/gutenberg/blob/5e30396ad24c7ea8c01030b04542daf62cb41de7/packages/components/src/context/use-context-system.js#L18-L29

https://github.com/WordPress/gutenberg/blob/5e30396ad24c7ea8c01030b04542daf62cb41de7/packages/components/src/context/utils.js#L10-L22


https://github.com/WordPress/gutenberg/blob/5e30396ad24c7ea8c01030b04542daf62cb41de7/packages/components/src/context/stories/ComponentsProvider.stories.js#L71-L72


Thanks in advance!
